File: 16566.py
from sys import stdin

input = stdin.readline


def solve():

    N, M, K = map(int, input().split())
    cards = sorted(map(int, input().split()))
    enemy_cards = [*map(int, input().split())]
    disjoint_set = [-1] * (M + 1)

    def union_find(a):
        if disjoint_set[a] == -1:
            return a
        disjoint_set[a] = union_find(disjoint_set[a])
        return disjoint_set[a]

    def bisect(x):
        lo, hi = 0, M
        while lo < hi:
            mid = (lo + hi) // 2
            if x < cards[union_find(mid)]:
                hi = mid
            else:
                lo = mid + 1

        return union_find(lo)

    for enemy in enemy_cards:
        idx = bisect(enemy)
        disjoint_set[idx] = union_find(idx+1)
        print(cards[idx])


if __name__ == '__main__':
    solve()

File: 16566.js
const fs = require('fs');

function solve() {
	let input = fs.readFileSync('/dev/stdin').toString().split('\n');
	let inputIdx = 0;

	const [N, M, K] = input[inputIdx++].split(' ').map((v) => +v);

	const cards = input[inputIdx++]
		.split(' ')
		.map((v) => +v)
		.sort((a, b) => a - b);

	const enemyCards = input[inputIdx++].split(' ').map((v) => +v);

	const disjointSet = [];
	disjointSet.length = M + 1;
	for (let i = 0; i < M + 1; disjointSet[i++] = -1);

	const union_find = (a) => {
		if (disjointSet[a] === -1) return a;
		disjointSet[a] = union_find(disjointSet[a]);
		return disjointSet[a];
	};

	const bisect = (x) => {
		let lo = 0,
			hi = M,
			mid;
		while (lo < hi) {
			mid = Math.floor((lo + hi) / 2);
			if (x < cards[union_find(mid)]) hi = mid;
			else lo = mid + 1;
		}
		return union_find(lo);
	};

	let ans = '';
	for (const enemy of enemyCards) {
		const idx = bisect(enemy);
		disjointSet[idx] = union_find(idx + 1);
		ans += `${cards[idx]}\n`;
	}

	console.log(ans);
}

solve();
